django/utils/decorators.py

```diff
@@ -17,13 +17,29 @@ def __get__(self, instance, owner):
         return super(classonlymethod, self).__get__(instance, owner)
 
 
-def method_decorator(decorator):
+def method_decorator(decorator, name=""):
     """
     Converts a function decorator into a method decorator
     """
     # 'func' is a function at the time it is passed to _dec, but will eventually
     # be a method of the class it is defined on.
-    def _dec(func):
+    # 'name' is required when a class is decorated instead of a method,
+    # and is the name of the method to be decorated, e.g. "dispatch".
+    def _dec(obj):
+        is_class = isinstance(obj, type)
+        if is_class:
+            if name and hasattr(obj, name):
+                func = getattr(obj, name)
```

Comment:
I've thought about it and tried some alternatives, but I think it more useful and correct to keep the two errors.

A `ValueError` can be caused to the *incorrect use of the keyword argument `name`*, by providing a value that is Falsy (the default) or does not correspond to a proper attribute on the class. Examples:
```python
@method_decorator(deco)
class MyView(View):
    ...
```
```python
@method_decorator(deco, name=None)
class MyView(View):
    ...
```
A `TypeError` occurs if `name` *is used correctly, but the resulting attribute isn't a callable object*. I think it's important to make the distinction.